File: lab-3/src/abstracts/Thing.java
package abstracts;

import com.company.Action;
import com.company.Quality;
import com.company.Sound;
import types.Color;

import java.util.Objects;

public abstract class Thing implements Paintable {

    protected String name;
    protected Sound sound;
    protected Quality quality;
    protected Color color;


    public Thing(String name, Sound sound, Quality quality, Color color) {
        this.name = name;
        this.sound = sound;
        this.quality = quality;
        this.color = color;
    }

    public void paint(Color color) {
        System.out.println(name + " теперь цвета " + color);
        this.color = color;
    }

    public void toSound() {
        System.out.print(name + ": ");
        sound.sound();
    }

    public Action toAppearOn(Thing thing) {
        return new Action(name + " цвета " + color + " появилось на " + thing.getName() +
                " " + "цвета " + thing.getColor(),
                new Quality("Появление"));
    }

    Action toAct(String name) {
        return new Action(name, quality);
    }

    public String getName() {
        return name;
    }

    public Sound getSound() {
        return sound;
    }

    public Quality getQuality() {
        return quality;
    }

    public Color getColor() {
        return color;
    }

    public void setColor(Color color) {
        this.color = color;
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Thing thing = (Thing) o;
        return name.equals(thing.name) && sound.equals(thing.sound) && quality.equals(thing.quality) && color == thing.color;
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, sound, quality, color);
    }

}

File: lab-4-kotlin/src/abstracts/Thing.kt
package abstracts

import abstracts.Thing
import abstracts.Closable
import abstracts.Alive
import kotlin.Throws
import exc.IncorrectFileNameException
import kotlin.jvm.JvmStatic
import exc.NogivenNameException
import com.company.things.Rocket
import com.company.things.Shots
import com.company.things.Bullets
import com.company.things.Something
import com.company.things.ShirtSleev
import com.company.things.Bloodstain
import com.company.things.Door
import com.company.things.Healpack
import types.Feeling
import java.lang.ArrayIndexOutOfBoundsException
import java.io.File
import java.io.FileNotFoundException
import java.util.HashSet
import com.company.Person.Face
import java.util.Objects
import abstracts.Paintable
import com.company.*
import types.Color

abstract class Thing(
    var name: String,
    var sound: Sound,
    var quality: Quality,
    var color: Color
) : Paintable {

    override fun paint(color: Color) {
        println("$name теперь цвета $color")
        this.color = color
    }

    fun toSound() {
        print("$name: ")
        sound.sound()
    }

    fun toAppearOn(thing: Thing): Action {
        return Action(
            name + " цвета " + color + " появилось на " + thing.name +
                    " " + "цвета " + thing.color,
            Quality("Появление")
        )
    }

    fun toAct(name: String): Action {
        return Action(name, quality)
    }

    override fun toString(): String {
        return name
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val thing = o as Thing
        return name == thing.name && sound == thing.sound && quality == thing.quality && color == thing.color
    }

    override fun hashCode(): Int {
        return Objects.hash(name, sound, quality, color)
    }
}